Add find_spec and exec_module to RestrictedModuleLoader

s-hertel opened the pull request:
##### SUMMARY
Update RestrictedModuleLoader for Python 3.10

Fixes #74658

##### ISSUE TYPE
- Feature Pull Request

s-hertel committed:
Add find_spec and exec_module to RestrictedModuleLoader

ssbarnea commented:
@s-hertel Any change to get this ready?

s-hertel committed:
Fix getting new loader with correct path

Fix pep8 errors

s-hertel committed:
Use convert_ansible_name_to_absolute_paths instead of the loader path

nitzmahone reviewed (approved):
Yep, this looks good.

The `exec_module` change for intercepting basic bugs me a bit- it feels like it should be executing on the module object passed in rather than doing the swap-a-dee-doo with the one loaded by another loader, but I get why it is the way it is, and I think it's safe enough. The only potentially simpler thing might be to pre-seed the modified basic to sys.modules and just remove the special cases from this loader entirely, but IIRC that gets frequently clobbered by this test by design, so this might just be the simplest way. Thanks!

EDIT: Matt C confirms that pre-seeding probably isn't worth the trouble; since this isn't a general-purpose loader, I think we can get away with the non-kosher swap behavior in `exec_module`, but we wouldn't want to do anything like that in, eg, the collection loader.

s-hertel commented on test/lib/ansible_test/_util/target/sanity/import/importer.py:
```diff
@@ -191,9 +221,36 @@ def find_module(self, fullname, path=None):
+            return None
+
+        def exec_module(self, module):
+            """Execute the module if the name is restricted and otherwise raise an ImportError
```
```suggestion
            """Execute the module if the name is ansible.module_utils.basic and otherwise raise an ImportError
```

s-hertel commented on test/lib/ansible_test/_util/target/sanity/import/importer.py:
```diff
@@ -191,9 +221,36 @@ def find_module(self, fullname, path=None):
+
         def load_module(self, fullname):
-            """Raise an ImportError.
+            """Return the module if the name is restricted and otherwise raise an ImportError.
```
```suggestion
            """Return the module if the name is ansible.module_utils.basic and otherwise raise an ImportError.
```

s-hertel committed:
Apply suggestions from code review

mattclay commented:
/azp run

mattclay commented on test/lib/ansible_test/_util/target/sanity/import/importer.py:
```diff
@@ -149,10 +157,32 @@ def __init__(self, path, name, restrict_to_module_paths):
 
+        def find_spec(self, fullname, path=None, target=None):  # pylint: disable=unused-argument
+            """Return the spec from the loader or None
+            :param fullname: str
+            :param path: List[str]
+            :param target: str
+            :return: ModuleSpec | None
```
The `:param` and `:return:` docstring entries here (and below) should be replaced comment based type hints.

The existing ones are also incorrect and should be type comments instead.

mattclay replied:
There are also quite a few `:type` type comments that were never converted to PEP 484 syntax. While those should also be fixed, I can take care of that in a separate PR once this is merged.

webknjaz replied:
Note that `:param:` entries should contain text descriptions, not types. Same for `:returns:`. There's a separate `:rtype:` for the type but with type annotation comments it won't be necessary.

s-hertel replied:
Fixed, and rebased this to test the type hints in CI too.

s-hertel replied:
Looks like CI found broken stuff though, so I'll need some help figuring out how to test these properly.

nitzmahone replied:
Hrm, that's weird- we've got blanket `ignore_missing_imports` for both `importlib.*` and `collections.*`, which should prevent the error you're getting. That's documented to also apply to subpackages, but maybe try adding `[mypy-collection.abc.*]` et al to that file in your PR to see if that fixes it?

https://github.com/ansible/ansible/commit/3d5637beecbbdcd1bc11a85d180de2087ecf12ee#diff-ed32086a44539250b26aa1f15420c7d79aad51f971a3126aece5ac419b3efdb5R91-R95

s-hertel replied:
That did fix it. The issue was caused by running the mypy test for this file against remote-only python versions (in addition to controller python versions). I simplified things in the last commit because I'm not sure whether or not it's helpful (the other mypy test targets in test/lib/ansible_test/_internal/ are only tested against controller python versions), but I can drop that commit removing that if it seems useful.

nitzmahone replied:
It's not a big deal either way, but since import sanity actually has to run on 2.x, it'd be *nice* if the type hint support was getting exercised there as well. Honestly I don't have strong feelings either way- I'm happy to merge as-is and we can go fix it and turn on remote-side stuff later (or not), or we can wait for Matt C to be back late next week and see what he says.

There are also a couple other ways to fix this for 2.x (but a bit gnarlier at the import sites). I suspect that having the blanket ignores present for 2.x's benefit could mask potential problems or missing typeshed stuff in 3.x, so we *might* want to look at the fully-inline options if/when we light this up for 2.x.

mattclay commented on test/lib/ansible_test/_util/target/sanity/import/importer.py:
```diff
@@ -149,10 +157,32 @@ def __init__(self, path, name, restrict_to_module_paths):
+            if loader is not None:
+                if has_py3_loader:
+                    return spec_from_loader(fullname, loader)
+                raise ImportError("Failed to import '%s' due to a bug in ansible-test. Check importlib imports for typos." % fullname)
```
I'm not clear on what conditions would trigger this, or what the user's action should be if it did occur.

s-hertel replied:
This was a safeguard against import issues in this file. If the py3 codepath is used, the py3 libs must be available, and if not it's probably a bug. I had misspelled SourceFileLoader in the import and it took me quite a while to find the actual issue. So I guess if it occurs, the user should open a bug with ansible/ansible to fix ansible-test. It might also occur for broken python3 installations, but I haven't tried to do that. Want me to change how this is handled?

mattclay commented on test/lib/ansible_test/_util/target/sanity/import/importer.py:
```diff
@@ -191,9 +221,36 @@ def find_module(self, fullname, path=None):
+                    real_module._load_params = lambda *args, **kwargs: {}  # pylint: disable=protected-access
+                    sys.modules[fullname] = real_module
+                    return
+            raise ImportError('import of "%s" is not allowed in this context' % fullname)
```
Just to be safe, we should raise an exception here if `ansible.module_utils.basic` was not found. It shouldn't happen, but if it does, the `ImportError` message will be wrong and confusing.

s-hertel replied:
I'm confused why this error should'nt be the same as the Py2 codepath. Is ImportError also wrong in that case?

Is it the type (ImportError) that's confusing here, or the error message?

nitzmahone replied:
I think he's suggesting a second error handler- currently this fall-through would be used for both "you tried to exec_module something that wasn't ansible.module_utils.basic" (for which the current message is fine), but also the case where the for loop completed without finding basic (in which case the error message will be incorrect and confusing). Basically just add an `else` clause to the end of the `for path in convert...` loop that raises a different ImportError that we couldn't find basic.

s-hertel replied:
Oh, I totally misunderstood that ask. That makes sense. Good catch.

ssbarnea commented:
I wonder if this might fix https://github.com/ansible-community/ansible-lint/issues/1919

s-hertel committed:
Fix type hints and test them in CI

s-hertel commented:
@ssbarnea I tested with the reproducer in your issue and confirmed that it was fixed by the PR linked from the issue comment https://github.com/ansible-community/ansible-lint/issues/1919#issuecomment-1063547436.

s-hertel committed:
Fix error message for ansible.module_utils.basic if it's missing

Add mypy ignored missing imports for controller sanity tests

s-hertel committed:
Add mypy attr-defined ignore entries for python 3.8, 3.9, 3.10 for vendored six

Add mypy attr-defined ignore for python 2.7 in lib/ansible/utils/collection_loader/_collection_finder.py

s-hertel committed:
Just test controller python versions to simplify ignoring mypy errors
